crawler/twitter.py

```diff
@@ -0,0 +1,183 @@
+from neo4j.v1 import GraphDatabase, basic_auth
+from TwitterAPI import TwitterAPI
+import json
+import os
+
+class TwitterCrawler:
+
+    GERMAN = "de";
+    USERS_SHOW = "users/show";
+    FOLLOWERS_IDS = "followers/ids";
+
+    def __init__(self):
+        self.driver = GraphDatabase.driver(
+            os.environ.get('neo4j_url'),
+            auth=(
+                os.environ.get('neo4j_username'),
+                os.environ.get('neo4j_password')
+            )
+        )
+        self.db = self.driver.session()
+        self.token = os.environ.get('APP_TOKEN')
+        self.secret = os.environ.get('APP_SECRET')
+
+    def start_crawling(self):
+        twitterHelpers = self.__get_twitter_helpers()
+        twitterHelper = twitterHelpers.pop()
+
+        for user in self.__get_twitter_users_queue():
+            response = self.__validate_twitter_user(twitterHelper, user["id"])
+            if response is True:
+                self.__save_twitter_followers(twitterHelper, user["id"])
+                self.__delete_twitter_user_from_queue(user["id"])
+            elif response == "Switch helper" or response == "Cannot authenticate":
+                if len(twitterHelpers) == 0:
+                    print("Starting recursive crawling.")
+                    self.start_crawling()
+                else:
+                    print("Switching twitter helper.")
+                    twitterHelper = twitterHelpers.pop()
+            elif response == "User not found" or response == "Suspended user":
+                self.__delete_twitter_user_from_queue(user["id"])
+                print("Deleting user id: " + user["id"] + " from users queue.")
+            elif response is not False:
+                print(response)
+
+    def __save_twitter_followers(self, twitterHelper, userId):
+        api = TwitterAPI(
+            self.token,
+            self.secret,
+            twitterHelper["accessToken"],
+            twitterHelper["accessTokenSecret"]
+        )
+
+        cursor = -1
+        while cursor == -1:
+            response = api.request(self.FOLLOWERS_IDS, {
+                "cursor": cursor,
+                "user_id": userId,
+                "count": 5000,
+                "stringify_ids": "true"
+            })
+
+            parsedResponse = json.loads(response.text)
```

Comment:
Here, when storing the followers of the userId, you only use information from Twitter. What would happen if a followship relation existed previously in the database but stopped existing on Twitter? Somehow, you should query the database to see who is following the userId for comparing database/Twitter data and potentially deleting relations that no longer exist... Right?

Reply:
done